File: python/help_handler.py
'''
MIT License

Copyright (c) 2022 TechnicFully
 
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''

from __future__ import print_function #For compatibility between Python 2 and 3 print functions
import os
import sys, re, warnings


#██╗   ██╗███████╗███████╗██████╗     ██╗   ██╗ █████╗ ██████╗ ██╗ █████╗ ██████╗ ██╗     ███████╗███████╗
#██║   ██║██╔════╝██╔════╝██╔══██╗    ██║   ██║██╔══██╗██╔══██╗██║██╔══██╗██╔══██╗██║     ██╔════╝██╔════╝
#██║   ██║███████╗█████╗  ██████╔╝    ██║   ██║███████║██████╔╝██║███████║██████╔╝██║     █████╗  ███████╗
#██║   ██║╚════██║██╔══╝  ██╔══██╗    ╚██╗ ██╔╝██╔══██║██╔══██╗██║██╔══██║██╔══██╗██║     ██╔══╝  ╚════██║
#╚██████╔╝███████║███████╗██║  ██║     ╚████╔╝ ██║  ██║██║  ██║██║██║  ██║██████╔╝███████╗███████╗███████║
# ╚═════╝ ╚══════╝╚══════╝╚═╝  ╚═╝      ╚═══╝  ╚═╝  ╚═╝╚═╝  ╚═╝╚═╝╚═╝  ╚═╝╚═════╝ ╚══════╝╚══════╝╚══════╝
#Magic numbers
found_all     = 0x0a4f781db
found_help    = 0x7b0acb244
found_version = 0xaa502bf21
found_none    = 0x819657e71




#██╗███╗   ██╗████████╗███████╗██████╗ ███╗   ██╗ █████╗ ██╗         ██╗   ██╗ █████╗ ██████╗ ██╗ █████╗ ██████╗ ██╗     ███████╗███████╗
#██║████╗  ██║╚══██╔══╝██╔════╝██╔══██╗████╗  ██║██╔══██╗██║         ██║   ██║██╔══██╗██╔══██╗██║██╔══██╗██╔══██╗██║     ██╔════╝██╔════╝
#██║██╔██╗ ██║   ██║   █████╗  ██████╔╝██╔██╗ ██║███████║██║         ██║   ██║███████║██████╔╝██║███████║██████╔╝██║     █████╗  ███████╗
#██║██║╚██╗██║   ██║   ██╔══╝  ██╔══██╗██║╚██╗██║██╔══██║██║         ╚██╗ ██╔╝██╔══██║██╔══██╗██║██╔══██║██╔══██╗██║     ██╔══╝  ╚════██║
#██║██║ ╚████║   ██║   ███████╗██║  ██║██║ ╚████║██║  ██║███████╗     ╚████╔╝ ██║  ██║██║  ██║██║██║  ██║██████╔╝███████╗███████╗███████║
#╚═╝╚═╝  ╚═══╝   ╚═╝   ╚══════╝╚═╝  ╚═╝╚═╝  ╚═══╝╚═╝  ╚═╝╚══════╝      ╚═══╝  ╚═╝  ╚═╝╚═╝  ╚═╝╚═╝╚═╝  ╚═╝╚═════╝ ╚══════╝╚══════╝╚══════╝                                                                                                                                    
_extraStringsGlob = True
_noArgHelpGlob = True
_matchHyphens = True
_hyphensOnly = False
_unknownArgHelpGlob = False
_disableOutput = False


_verGlob = "No version available"
_appNameGlob = ""




class HelpHandler():
    #██████╗ ██████╗ ██╗██╗   ██╗ █████╗ ████████╗███████╗    ███╗   ███╗███████╗████████╗██╗  ██╗ ██████╗ ██████╗ ███████╗
    #██╔══██╗██╔══██╗██║██║   ██║██╔══██╗╚══██╔══╝██╔════╝    ████╗ ████║██╔════╝╚══██╔══╝██║  ██║██╔═══██╗██╔══██╗██╔════╝
    #██████╔╝██████╔╝██║██║   ██║███████║   ██║   █████╗      ██╔████╔██║█████╗     ██║   ███████║██║   ██║██║  ██║███████╗
    #██╔═══╝ ██╔══██╗██║╚██╗ ██╔╝██╔══██║   ██║   ██╔══╝      ██║╚██╔╝██║██╔══╝     ██║   ██╔══██║██║   ██║██║  ██║╚════██║
    #██║     ██║  ██║██║ ╚████╔╝ ██║  ██║   ██║   ███████╗    ██║ ╚═╝ ██║███████╗   ██║   ██║  ██║╚██████╔╝██████╔╝███████║
    #╚═╝     ╚═╝  ╚═╝╚═╝  ╚═══╝  ╚═╝  ╚═╝   ╚═╝   ╚══════╝    ╚═╝     ╚═╝╚══════╝   ╚═╝   ╚═╝  ╚═╝ ╚═════╝ ╚═════╝ ╚══════╝
    @staticmethod
    def __print(string, newline=True):
        global _disableOutput
        if _disableOutput != True:
            if newline == True:
                print(string)
            else:
                print(string, end='')


    #██████╗ ██╗   ██╗██████╗ ██╗     ██╗ ██████╗    ███╗   ███╗███████╗████████╗██╗  ██╗ ██████╗ ██████╗ ███████╗
    #██╔══██╗██║   ██║██╔══██╗██║     ██║██╔════╝    ████╗ ████║██╔════╝╚══██╔══╝██║  ██║██╔═══██╗██╔══██╗██╔════╝
    #██████╔╝██║   ██║██████╔╝██║     ██║██║         ██╔████╔██║█████╗     ██║   ███████║██║   ██║██║  ██║███████╗
    #██╔═══╝ ██║   ██║██╔══██╗██║     ██║██║         ██║╚██╔╝██║██╔══╝     ██║   ██╔══██║██║   ██║██║  ██║╚════██║
    #██║     ╚██████╔╝██████╔╝███████╗██║╚██████╗    ██║ ╚═╝ ██║███████╗   ██║   ██║  ██║╚██████╔╝██████╔╝███████║
    #╚═╝      ╚═════╝ ╚═════╝ ╚══════╝╚═╝ ╚═════╝    ╚═╝     ╚═╝╚══════╝   ╚═╝   ╚═╝  ╚═╝ ╚═════╝ ╚═════╝ ╚══════╝

    ''' For configuring functionality that might conflict/clutter other program output. The parameters are as follows...
          no_arg_help         - Print help dialogue when no arguments are given
          extra_strings       - Whether to match for h, -h, --h, v, -v and --v specifically (which may conflict with your program’s flags)
          match_hyphens       - Match arguments beginning with hyphens (i.e., "help" vs "--help")
          hyphens_only        - Only match arguments that begin with one or more hyphens
          unknown_arg_help    - Print help dialogue when an unknown argument is passed. You would typically whitelist your program’s option flags in combination with this
          disable_output      - Disable all output of HelpHandler
    '''
    @staticmethod
    def config(no_arg_help=True, extra_strings=True, match_hyphens=True, hyphens_only=False, unknown_arg_help=False, disable_output=False):
        #Error checking
        if not (isinstance(no_arg_help, int)):
            raise TypeError("argument no_arg_help is not of type bool or int")
        if not (isinstance(extra_strings, int)):
            raise TypeError("argument extra_strings is not of type bool or int")
        if not (isinstance(match_hyphens, int)):
            raise TypeError("argument match_hyphens is not of type bool or int")
        if not (isinstance(hyphens_only, int)):
            raise TypeError("argument hyphens_only is not of type bool or int")
        if not (isinstance(unknown_arg_help, int)):
            raise TypeError("argument unknown_arg_help is not of type bool or int")
        if not (isinstance(disable_output, int)):
            raise TypeError("argument disable_output is not of type bool or int")


        global _extraStringsGlob
        global _noArgHelpGlob
        global _unknownArgHelpGlob
        global _matchHyphens
        global _hyphensOnly
        global _disableOutput

        _extraStringsGlob   = extra_strings
        _noArgHelpGlob      = no_arg_help
        _matchHyphens       = match_hyphens
        _hyphensOnly        = hyphens_only
        _unknownArgHelpGlob = unknown_arg_help
        _disableOutput      = disable_output


    # Set your program version which will be output as appropriate. This shouldn't be anything fancy, just a simple version number
    # Throws an error if the given version is not a str / basestring, int, or float type
    @staticmethod
    def version(version):
        #Error checking
        if sys.version_info.major >= 3:
            if not (isinstance(version, str), isinstance(version, int), isinstance(version, float)):
                raise TypeError("argument version is not of type string, int, or double")
        elif sys.version_info.major <= 2:
            if not (isinstance(version, basestring), isinstance(version, int), isinstance(version, float)):
                raise TypeError("argument version is not of type string, int, or double")

        if isinstance(version, list) or isinstance(version, tuple) or isinstance(version, dict):
            raise TypeError("argument version is not of type string, int, or double")

        if type(version) == type(True):
            raise TypeError("argument version should not be of type boolean")

        if not version:
            raise ValueError("given version string is empty")


        global _verGlob
        _verGlob = version

    # Defines your program name which will be output alongside help dialogue
    # Throws an error if the given app_name is not a string, or empty
    @staticmethod
    def name(app_name):
        #Error checking
        if sys.version_info.major >= 3:
            if not (isinstance(app_name, str)):
                raise TypeError("argument app name is not of type string")
        elif sys.version_info.major <= 2:
            if not (isinstance(app_name, basestring)):
                raise TypeError("argument app name is not of type basestring")

        if not app_name:
            raise ValueError("given app name string is empty")
        

        global _appNameGlob
        _appNameGlob = app_name

    # A single function for passing your program's name as well as its version
    # Returns the same values as HelpHandler.name() and HelpHandler.version()
    # Throws the same errors as HelpHandler.name() and HelpHandler.version()
    @staticmethod
    def info(app_name, version):
        HelpHandler.name(app_name)
        HelpHandler.version(version)

    # This is the main function which processes and outputs the appropriate dialogue based on the user's input. You must pass or set any other options and info before calling this
    # Returns found_all, found_help, found_version, or found_none
    @staticmethod
    def handle(help_dialogue):
        global _disableOutput
        
        if not help_dialogue:
            help_dialogue = "No usage help is available"
        if (len(sys.argv) == 1) and (_noArgHelpGlob == True):
            HelpHandler.__print(help_dialogue)
            return found_none
        
        #Set regex
        numHyphens = 0
        if _matchHyphens == False:
            numHyphens = 1
        numHyphens = str(numHyphens)

        helpRegex    = "-{" + numHyphens + ",}h{1,}e{1,}l{1,}p{1,}(.*)"
        versionRegex = "-{" + numHyphens + ",}v{1,}e{1,}r{1,}s{1,}i{1,}o{1,}n{1,}(.*)"

        global _extraStringsGlob
        if _extraStringsGlob == True:
            helpRegex    += "|-{" + numHyphens + ",}h{1,}$"
            versionRegex += "|^-{" + numHyphens + ",}v$"
            

        #Match and count arguments
        foundMatchHelp = False
        foundMatchVer  = False
        matches = 0
        for arg in sys.argv[1:]:
            if re.match(versionRegex, arg, flags=re.IGNORECASE):
                foundMatchVer = True
                matches += 1
        for arg in sys.argv[1:]:
            if re.match(helpRegex, arg, flags=re.IGNORECASE):
                foundMatchHelp = True
                matches += 1

        #Print respective output if arguments found and return number of args matched
        if _disableOutput != True:
            global _appNameGlob
            global _verGlob
            if foundMatchHelp and foundMatchVer:
                if _appNameGlob:
                    HelpHandler.__print(_appNameGlob + " ", False)

                HelpHandler.__print(_verGlob)
                HelpHandler.__print(help_dialogue)
            elif foundMatchHelp == True:
                if _appNameGlob:
                    HelpHandler.__print(_appNameGlob + " ", False)
                HelpHandler.__print(help_dialogue)
            elif foundMatchVer == True:
                HelpHandler.__print(_verGlob)


        #Return
        if foundMatchHelp and foundMatchVer:
            return found_all
        if foundMatchHelp:
            return found_help
        elif foundMatchVer:
            return found_version
        else:
            #Nothing matched
            if _unknownArgHelpGlob == True and _disableOutput != True:
                if (len(sys.argv) > 2):
                    HelpHandler.__print("Unknown arguments given")
                else:
                    HelpHandler.__print("Unknown argument given")

            return found_none

    # This function like helpHandler.handle, will processes and output the appropriate dialogue based on the user's input, but using a file as its dialogue source. You must pass or set any other options and info before calling this
    # Throws an error if the given file_name could not be opened (does not exist, etc.)
    @staticmethod
    def handleFile(file_name):
        if not os.path.exists(file_name):
            if sys.version_info.major >= 3:
                raise FileNotFoundError("given file name does not refer to an existing file")
            elif sys.version_info.major <= 2:
                raise IOError("given file name does not refer to an existing file")

        f = open(file_name, 'r')
        contents = f.read()

        if re.search(r'^\s*$', contents):
            f.close()
            raise EOFError("given file name was found, but contains no data")

        f.close()
        return HelpHandler.handle(contents)

File: java/HelpHandler.java
/* MIT License
 *
 * Copyright (c) 2022 TechnicFully
 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */


import java.util.regex.*;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.Path;
import java.io.IOException;




public class HelpHandler {
    final static public int helpHandlerNone   = 0; //Returned if no arguments were matched; otherwise the number of arguments matched are returned

    private static boolean optNoArgHelp = true;
    private static boolean optExtraStrings = true;
    private static boolean optUnknownArgHelp = false;

    
    private static String  verStr = "No version is available";
    private static double  verNum = 0;
    private static String  name = "";

    
    final private static int typeVerStr = 0;
    final private static int typeVerNum = 1;
    private static  int mostRecentVer = typeVerStr;



    static String readFile(String path, Charset encoding) throws IOException {
        byte[] encoded = Files.readAllBytes(Paths.get(path));
        if (encoded.length <= 0) {
            throw new IOException("usage file is empty");
        }

        return new String(encoded, encoding);
    }

    //Set your programs version which will be output as appropriate. This shouldn't be anything fancy, just a simple version number
    public static void version(final String version) throws IllegalArgumentException {
        if (version == null) {
            throw new IllegalArgumentException("version string is null"); }
        if (version.length() <= 0) {
            throw new IllegalArgumentException("version string is empty"); }

        verStr = version;
        mostRecentVer = typeVerStr;
    }
    //Set your programs version which will be output as appropriate. This shouldn't be anything fancy, just a simple version number
    public static void version(final Double version) {
        verNum = version;
        mostRecentVer = typeVerNum;
    }

    //Defines your program name which will be output alongside help dialogue
    public static void name(final String appName) throws IllegalArgumentException {
        if (appName == null) {
            throw new IllegalArgumentException("app name string is null"); }
        if (appName.length() <= 0) {
            throw new IllegalArgumentException("app name string is empty"); }

        name = appName;
    }

    //A single function for passing your program's name as well as its version
    public static void info(final String appName, final String version) throws IllegalArgumentException {
        HelpHandler.name(appName);
        HelpHandler.version(version);
    }
    //A single function for passing your program's name as well as its version
    public static void info(final String appName, final Double version) throws IllegalArgumentException {
        HelpHandler.name(appName);
        HelpHandler.version(version);
    }

    //For configuring functionality that might conflict/clutter other program output
    public static void config(final boolean noArgHelp) {
        optNoArgHelp = noArgHelp;

        return;
    }
    //For configuring functionality that might conflict/clutter other program output
    public static void config(final boolean noArgHelp, final boolean extraStrings) {
        config(noArgHelp);
        optExtraStrings = extraStrings;

        return;
    }
    //For configuring functionality that might conflict/clutter other program output
    public static void config(final boolean noArgHelp, final boolean extraStrings, final boolean unknownArgHelp) {
        config(noArgHelp, extraStrings);
        optUnknownArgHelp = unknownArgHelp;

        return;
    }

    //This is the main function which processes and outputs the appropriate dialogue based on the user's input. You must pass or set any other options and info before calling this
    public static int handle(final String args[], String helpDialogue) throws RuntimeException {
        if (helpDialogue.length() <= 0)
            helpDialogue = "No usage help is available";
        if (args.length == 0 && optNoArgHelp == true ) {
            System.out.println(helpDialogue);
            System.out.flush();
            return helpHandlerNone; }


        //Error checks
        if (args == null) {
            throw new RuntimeException("argument vector is null"); }
        if (args.length < 0) {
            throw new RuntimeException("argument vector count is below 0 (should always be at least 0)"); }


        //Set regex
        String regexHelp = "-{0,}h{1,}e{1,}l{1,}p{1,}(.*)";
        String regexVer  = "-{0,}v{1,}e{1,}r{1,}s{1,}i{1,}o{1,}n{1,}(.*)";

        if (optExtraStrings == true) {
            regexHelp += "|-{0,}h{1,}$";
            regexVer  += "|^-{0,}v$";
        }


        //Match and count arguments
        boolean matchedHelp = false;
        boolean matchedVer = false;
        int matches = 0;
        for (int i = 0; i < args.length; i++) {
            if (args[i] == null) {
                throw new RuntimeException("argument index " + i + " is null despite being of length greater than 0"); }
            if (Pattern.matches(regexHelp, args[i].toLowerCase())) { //Might switch to setting the CASE_INSENSITIVE regex flag as it seems to be ~50ns faster
                matchedHelp = true;
                matches++; } 
            if (Pattern.matches(regexVer, args[i].toLowerCase())) {
                matchedVer = true;
                matches++; } 
        }


        //Print respective output if arguments found and return number of args matched
        if (matchedHelp && matchedVer) {
            if (name.length() > 0) {
                System.out.print(name + " ");
            }
            
            System.out.println(verStr);
            System.out.println(helpDialogue);
        } else if (matchedVer == true) {
            if (mostRecentVer == typeVerStr) {
                System.out.println(verStr);
            } else if (mostRecentVer == typeVerNum) {
                System.out.println(verNum); }
            System.out.flush(); 
        } else if (matchedHelp == true) {
            if (name.length() > 0) {
                System.out.print(name + " ");
            }

            System.out.println(helpDialogue);
            System.out.flush(); }

        if (matchedVer == true || matchedHelp == true) {
            return matches; 
        }


        //Nothing matched
        if (optUnknownArgHelp == true) {
            String result = (args.length > 2) ? "Unknown arguments given" : "Unknown argument given";
            System.out.println(result);
            System.out.flush();
        }

        
        return helpHandlerNone;
    }

    //This function like helpHandler.handle, will processes and output the appropriate dialogue based on the user's input, but using a file as its dialogue source. You must pass or set any other options and info before calling this
    public static int handleFile(final String args[], final String file_name) throws Exception {
        if (file_name == null) {
            throw new IllegalArgumentException("file name is null");
        }
        if (file_name.length() <= 0) {
            throw new IllegalArgumentException("file name is empty");
        }

        Path path = Paths.get(file_name);
        if (Files.notExists(path)) {
            throw new IOException("file name does not refer to an existing file");
        }

        String contents = readFile(file_name, Charset.defaultCharset());
        return HelpHandler.handle(args, contents);
    }

}
